frontend/services/viewer/viewerService/viewerController.tsx

```diff
@@ -0,0 +1,666 @@
+			}
+		}
+	}
+
+	/**
+	 * Measure
+	 */
+
+	public async activateMeasure() {
+		await this.isViewerReady();
+		this.setMeasureMode(true);
+	}
+
+	public async disableMeasure() {
+		await this.isViewerReady();
+		this.setMeasureMode(false);
+	}
+
+	public setMeasureMode(on: boolean) {
+		this.measureMode = on;
+		if (on === true) {
+			UnityUtil.enableMeasuringTool();
+		} else {
+			UnityUtil.disableMeasuringTool();
+		}
+	}
+
+	/**
+	 * Highlight
+	 */
+
+	public get canHighlight() {
+		return this.initialized && !this.pinDropMode && !this.measureMode;
+	}
+
+	public async highlightObjects(params) {
+		if (this.canHighlight) {
+			const ids = params.id ? [params.id] : params.ids;
+			if (ids) {
+				const uniqueIds = Array.from(new Set(ids));
+				if (uniqueIds.length) {
+					await this.highlightAllObjects(
+							params.account,
+							params.model,
+							ids,
+							params.colour,
+							params.multi,
+							params.forceReHighlightt);
+
+					this.emit(VIEWER_EVENTS.HIGHLIGHT_OBJECTS, {account: params.account, model: params.model, uniqueIds });
+					return;
+				}
+			}
+
+			this.clearHighlights();
+		}
+	}
+
+	public unhighlightObjects(params) {
+		this.unhighlightAllObjects(
+				params.account,
+				params.model,
+				params.id ? [params.id] : params.ids
+		);
+	}
+
+	/**
+	 * Settings
+	 */
+
+	public updateViewerSettings(settings) {
+		if (settings) {
+			this.settings = settings;
+			if (this.settings.properties && this.settings.properties.unit) {
+				this.setUnits(this.settings.properties.unit);
+			}
+		}
+	}
+
+	public setUnits(units) {
+		this.units = units;
+
+		if (units === 'mm') {
+			this.convertToM = 0.001;
+		} else if (units === 'ft') {
+			this.convertToM = 0.0032;
+		}
+
+		// Set the units in unity for the measure tool
+		if (this.units) {
+			UnityUtil.setUnits(this.units);
+		}
+	}
+
+	/**
+	 * Nav mode
+	 */
+
+	public setNavigationMode(mode) {
+		this.setNavMode(mode, false);
+	}
+
+	public setNavMode(mode, force) {
+		if (this.currentNavMode !== mode || force) {
+			this.currentNavMode = mode;
+			this.setNavigation(mode);
+		}
+	}
+
+	public navMethodChanged(newNavMode) {
+		this.currentNavMode = newNavMode;
+	}
+
+	public get getNavMode() {
+		return this.currentNavMode;
+	}
+
+	/**
+	 * Fullscreen
+	 */
+
+	public switchFullScreen() {
+		if (!this.fullscreen) {
+			if (this.viewer.hasOwnProperty('mozRequestFullScreen')) {
+			} else if (this.viewer.webkitRequestFullscreen) {
+				this.viewer.webkitRequestFullscreen();
+			}
+
+			this.fullscreen = true;
+		} else {
+			if (document.webkitCancelFullScreen) {
+				document.webkitCancelFullScreen();
+			}
+
+			this.fullscreen = false;
+		}
+	}
+
+	/**
+	 * Pins
+	 */
+
+	public addPin = async ({account, model, id, type, pickedPos: position, norm, colours, viewpoint, isSelected}) => {
+			await this.isViewerReady();
+			this.pins[id] = new Pin({ id, type, position, norm, colors: colours, account, model });
```

Comment:
I don't think the pin file needs to exist, just pass the info straight to unity. Same with any oother pin functions here that calls this.pins...
There's not a lot of point having an instance of the pin in the viewer unless it's meaningful, and the state of the pins seems to be captured outside of this anyway.

@sanmont3drepo do you agree?

Reply:
I agree, I saw that when i was fixing the pin bug and I had the same hunch.